File: services/go-gateway/internal/middleware/metrics.go
// Package middleware contem middlewares HTTP para o gateway.
package middleware

import (
	"log/slog"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

// RequestLogger loga entrada e saida de cada request com duracao em ms.
// Espera que o middleware RequestID() tenha sido executado antes.
func RequestLogger() gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()

		requestID, _ := c.Get("request_id")

		slog.Info("Request started",
			slog.String("request_id", requestID.(string)),
			slog.String("method", c.Request.Method),
			slog.String("path", c.Request.URL.Path),
		)

		c.Next()

		durationMs := float64(time.Since(start).Milliseconds())
		statusCode := c.Writer.Status()

		slog.Info("Request completed",
			slog.String("request_id", requestID.(string)),
			slog.String("method", c.Request.Method),
			slog.String("path", c.Request.URL.Path),
			slog.Int("status_code", statusCode),
			slog.Float64("duration_ms", durationMs),
		)
	}
}

// Metricas Prometheus coletadas pelo middleware MetricsCollector.
var (
	requestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "requests_total",
			Help: "Total number of HTTP requests",
		},
		[]string{"method", "path", "status_code"},
	)

	requestDuration = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "requests_duration_seconds",
			Help:    "HTTP request duration in seconds",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"method", "path", "status_code"},
	)
)

// MetricsCollector coleta latencia e contagem de requests para Prometheus.
func MetricsCollector() gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()

		c.Next()

		duration := time.Since(start).Seconds()
		statusCode := strconv.Itoa(c.Writer.Status())
		path := c.Request.URL.Path

		// Ignorar rota de metricas e health para evitar ruido nas metricas
		if path == "/metrics" || path == "/api/v1/health" || path == "/api/v1/ready" {
			return
		}

		requestsTotal.WithLabelValues(c.Request.Method, path, statusCode).Inc()
		requestDuration.WithLabelValues(c.Request.Method, path, statusCode).Observe(duration)
	}
}

File: services/python-agent/app/core/metrics.py
"""
Prometheus metrics for the Python agent service.

Exposes a /metrics endpoint and provides middleware that tracks:
- requests_total (counter with method, path, status_code labels)
- requests_duration_seconds (histogram with method, path, status_code labels)
"""

from __future__ import annotations

import time

from prometheus_client import Counter, Histogram
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response

# Contador total de requisicoes
requests_total = Counter(
    "requests_total",
    "Total number of HTTP requests",
    ["method", "path", "status_code"],
)

# Histograma de duracao das requisicoes em segundos
requests_duration = Histogram(
    "requests_duration_seconds",
    "HTTP request duration in seconds",
    ["method", "path", "status_code"],
    buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0),
)


class MetricsMiddleware(BaseHTTPMiddleware):
    """Middleware que coleta metricas Prometheus para cada requisicao."""

    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        start_time = time.perf_counter()

        response = await call_next(request)

        duration = time.perf_counter() - start_time
        status_code = str(response.status_code)
        path = request.url.path

        # Rotas de metricas e health nao devem ser contabilizadas
        if path in ("/metrics", "/api/v1/health", "/api/v1/ready"):
            return response

        requests_total.labels(
            method=request.method,
            path=path,
            status_code=status_code,
        ).inc()

        requests_duration.labels(
            method=request.method,
            path=path,
            status_code=status_code,
        ).observe(duration)

        return response
